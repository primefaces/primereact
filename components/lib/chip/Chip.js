import * as React from 'react';
import { classNames, IconUtils, ObjectUtils } from '../utils/Utils';

export const Chip = React.memo(React.forwardRef((props, ref) => {
    const [visibleState, setVisibleState] = React.useState(true);

    const onKeyDown = (event) => {
        if (event.keyCode === 13) { // enter
            close(event);
        }
    }

    const close = (event) => {
        setVisibleState(false);

        if (props.onRemove) {
            props.onRemove(event);
        }
    }

    const createContent = () => {
        let content = [];

        if (props.image) {
            content.push(<img key="image" src={props.image} alt={props.imageAlt} onError={props.onImageError}></img>);
        }
        else if (props.icon) {
            content.push(IconUtils.getJSXIcon(props.icon, { key: 'icon', className: 'p-chip-icon' }, { props }));
        }

        if (props.label) {
            content.push(<span key="label" className="p-chip-text">{props.label}</span>);
        }

        if (props.removable) {
            content.push(IconUtils.getJSXIcon(props.removeIcon, { key: 'removeIcon', tabIndex: 0, className: 'p-chip-remove-icon', onClick: close, onKeyDown }, { props }));
        }

        return content;
    }

    const createElement = () => {
        const otherProps = ObjectUtils.findDiffKeys(props, Chip.defaultProps);
        const className = classNames('p-chip p-component', {
            'p-chip-image': props.image != null
        }, props.className);

        const content = props.template ? ObjectUtils.getJSXElement(props.template, props) : createContent();

        return (
            <div className={className} style={props.style} {...otherProps}>
                {content}
            </div>
        )
    }

    return visibleState && createElement();
}));

Chip.displayName = 'Chip';
Chip.defaultProps = {
    __TYPE: 'Chip',
    label: null,
    icon: null,
    image: null,
    removable: false,
    removeIcon: 'pi pi-times-circle',
    className: null,
    style: null,
    template: null,
    imageAlt: 'chip',
    onImageError: null,
    onRemove: null
}
