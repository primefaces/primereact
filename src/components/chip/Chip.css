.p-chip {
    display: inline-flex;
    align-items: center;
}

.p-chip-text {
    line-height: 1.5;
}

.p-chip-icon.pi {
    line-height: 1.5;
}

.p-chip .p-chip-remove-icon {
    line-height: 1.5;
    cursor: pointer;
}

.p-chip img {
    border-radius: 50%;
}
