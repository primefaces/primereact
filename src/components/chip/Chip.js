import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { classNames } from '../utils/ClassNames';
import ObjectUtils from '../utils/ObjectUtils';

export class Chip extends Component {

    static defaultProps = {
        label: null,
        icon: null,
        image: null,
        removable: false,
        removeIcon: 'pi pi-times-circle',
        className: null,
        style: null,
        template: null,
        imageAlt: 'chip',
        onImageError: null,
        onRemove: null
    };

    static propTypes = {
        label: PropTypes.string,
        icon: PropTypes.string,
        image: PropTypes.string,
        removable: PropTypes.bool,
        removeIcon: PropTypes.string,
        className: PropTypes.string,
        style: PropTypes.object,
        template: PropTypes.any,
        imageAlt: PropTypes.string,
        onImageError: PropTypes.func,
        onRemove: PropTypes.func
    };

    constructor(props) {
        super(props);

        this.state = {
            visible: true
        };

        this.close = this.close.bind(this);
        this.onKeyDown = this.onKeyDown.bind(this);
    }

    onKeyDown(event) {
        if (event.keyCode === 13) { // enter
            this.close(event);
        }
    }

    close(event) {
        event.persist();

        this.setState({ visible: false }, () => {
            if (this.props.onRemove) {
                this.props.onRemove(event);
            }
        });
    }

    renderContent() {
        let content = [];

        if (this.props.image) {
            const onError = (e) => {
                if (this.props.onImageError) {
                    this.props.onImageError(e);
                }
            };

            content.push(<img key="image" src={this.props.image} alt={this.props.imageAlt} onError={onError}></img>);
        }
        else if (this.props.icon) {
            const iconClassName = classNames('p-chip-icon', this.props.icon);
            content.push(<span key="icon" className={iconClassName}></span>);
        }

        if (this.props.label) {
            content.push(<span key="label" className="p-chip-text">{this.props.label}</span>);
        }

        if (this.props.removable) {
            const removableIconClassName = classNames('p-chip-remove-icon', this.props.removeIcon);
            content.push(<span key="removeIcon" tabIndex={0} className={removableIconClassName} onClick={this.close} onKeyDown={this.onKeyDown}></span>);
        }

        return content;
    }

    renderElement() {
        const containerClassName = classNames('p-chip p-component', {
            'p-chip-image': this.props.image != null
        }, this.props.className);

        const content = this.props.template ? ObjectUtils.getJSXElement(this.props.template, this.props) : this.renderContent();

        return (
            <div className={containerClassName} style={this.props.style}>
                {content}
            </div>
        );
    }

    render() {
        return this.state.visible && this.renderElement();
    }
}
