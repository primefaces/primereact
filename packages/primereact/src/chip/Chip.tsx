'use client';
import { Component, withComponent } from '@primereact/core/component';
import { useChip } from '@primereact/headless/chip';
import { TimesCircleIcon } from '@primereact/icons/timescircle';
import { styles } from '@primereact/styles/chip';
import { mergeProps } from '@primeuix/utils';
import * as React from 'react';
import { defaultProps } from './Chip.props';
import { ChipIcon } from './icon';
import { ChipImage } from './image';
import { ChipLabel } from './label';

export const Chip = withComponent({
    defaultProps,
    styles,
    setup: (instance) => {
        const chip = useChip(instance.inProps);

        return chip;
    },
    render: (instance) => {
        const {
            id,
            props,
            ptmi,
            ptm,
            cx,
            // element refs
            elementRef,
            // methods
            onKeyDown,
            close,
            visibleState
        } = instance;

        const rootProps = mergeProps(
            {
                id,
                className: cx('root')
            },
            ptmi('root')
        );

        const createRemoveElement = () => {
            const removeIconProps = mergeProps(
                {
                    className: cx('removeIcon'),
                    tabIndex: 0,
                    onClick: close,
                    onKeyDown
                },
                ptm('removeIcon')
            );

            return <TimesCircleIcon {...removeIconProps} />;
        };

        return visibleState ? (
            <Component as={props.as || 'div'} {...rootProps} ref={elementRef}>
                {props.children}
                {props.removable && createRemoveElement()}
            </Component>
        ) : null;
    },
    components: {
        Icon: ChipIcon,
        Image: ChipImage,
        Label: ChipLabel
    }
});
