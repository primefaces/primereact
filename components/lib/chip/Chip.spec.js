import '@testing-library/jest-dom';
import { fireEvent, render } from '@testing-library/react';
import { Chip } from './Chip';

import { snapshot } from '../../test';

describe('Chip', () => {
    snapshot(<Chip />, 'default');
    snapshot(<Chip image="http://google.com/my.png" />, 'image');
    snapshot(<Chip image="http://google.com/my.png" imageAlt="jest" />, 'image and alt');
    snapshot(<Chip icon="pi pi-check" />, 'icon');
    snapshot(<Chip label="jest" />, 'label');
    test('when removable is true it returns with remove icon', () => {
        // Arrange
        const removeOn = jest.fn();
        const { container } = render(<Chip removable onRemove={removeOn} />);

        expect(container).toMatchSnapshot('before remove');
        const chipRemoveIcon = container.getElementsByClassName('p-chip-remove-icon')[0];

        // Act
        fireEvent.keyDown(chipRemoveIcon, { key: 'enter', keyCode: 13 });

        // Assert
        expect(container).toMatchSnapshot('after remove');
        expect(removeOn).toHaveBeenCalledTimes(1);
    });
    test('when removable is true the chip is removed when ENTER is pressed', () => {
        // Arrange
        const removeOn = jest.fn();
        const { container } = render(<Chip removable onRemove={removeOn} />);

        expect(container).toMatchSnapshot('before remove');
        const chipRemoveIcon = container.getElementsByClassName('p-chip-remove-icon')[0];

        // Act
        fireEvent.click(chipRemoveIcon);

        // Assert
        expect(container).toMatchSnapshot('after remove');
        expect(removeOn).toHaveBeenCalledTimes(1);
    });
});
