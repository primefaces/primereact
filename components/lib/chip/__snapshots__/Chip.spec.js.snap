// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chip default 1`] = `
<div>
  <div
    class="p-chip p-component"
  />
</div>
`;

exports[`Chip icon 1`] = `
<div>
  <div
    class="p-chip p-component"
  >
    <span
      class="p-chip-icon pi pi-check"
    />
  </div>
</div>
`;

exports[`Chip image 1`] = `
<div>
  <div
    class="p-chip p-component p-chip-image"
  >
    <img
      alt="chip"
      src="http://google.com/my.png"
    />
  </div>
</div>
`;

exports[`Chip image and alt 1`] = `
<div>
  <div
    class="p-chip p-component p-chip-image"
  >
    <img
      alt="jest"
      src="http://google.com/my.png"
    />
  </div>
</div>
`;

exports[`Chip label 1`] = `
<div>
  <div
    class="p-chip p-component"
  >
    <span
      class="p-chip-text"
    >
      jest
    </span>
  </div>
</div>
`;

exports[`Chip when removable is true it returns with remove icon: after remove 1`] = `<div />`;

exports[`Chip when removable is true it returns with remove icon: before remove 1`] = `
<div>
  <div
    class="p-chip p-component"
  >
    <span
      class="p-chip-remove-icon pi pi-times-circle"
      tabindex="0"
    />
  </div>
</div>
`;

exports[`Chip when removable is true the chip is removed when ENTER is pressed: after remove 1`] = `<div />`;

exports[`Chip when removable is true the chip is removed when ENTER is pressed: before remove 1`] = `
<div>
  <div
    class="p-chip p-component"
  >
    <span
      class="p-chip-remove-icon pi pi-times-circle"
      tabindex="0"
    />
  </div>
</div>
`;
